test_agent/rust/src/testagent.rs

```diff
@@ -0,0 +1,294 @@
+        let socket = Arc::new(Mutex::new(test_clientsocket));
+        let socket_to_tm = Arc::new(Mutex::new(test_clientsocket_to_tm));
+        let clientsocket = socket;
+        let clientsocket_to_tm = socket_to_tm;
+        SocketTestAgent {
+            utransport,
+            clientsocket,
+            clientsocket_to_tm,
+            listner_map: Vec::new(),
+        }
+    }
+
+
+    fn sanitize_input_string(self, input: &str) -> String {
+        input.chars()
+            .map(|c| {
+                match c {
+                    '\x00'..='\x1F' => self.clone().escape_control_character(c),
+                    _ => c.to_string(),
+                }
+            })
+            .collect()
+    }
+     
+    fn escape_control_character(self,c: char) -> String {
+        let escaped = format!("\\u{:04x}", c as u32);
+        escaped
+    }
+    pub async fn receive_from_tm(&mut self) {
+        // Clone Arc to capture it in the closure
+
+        let arc_self = Arc::new(self.clone());
+        self.clone().inform_tm_ta_starting();
+        let mut socket = self.clientsocket.lock().expect("error accessing TM server");
+
+        loop {
+            let mut recv_data = [0; 2048];
+
+            let bytes_received = match socket.read(&mut recv_data) {
+                Ok(bytes_received) => bytes_received,
+                Err(e) => {
+                    // Handle socket errors (e.g., connection closed)
+                    dbg!("Socket error: {}", e);
+                    break;
+                }
+            };
+            // Check if no data is received
+            if bytes_received == 0 {
+                continue;
+            }
+
+            let recv_data_str: std::borrow::Cow<'_, str> =
+            String::from_utf8_lossy(&recv_data[..bytes_received]);
+            let mut action_str = "";
+            let cleaned_json_string = self.clone().sanitize_input_string(&recv_data_str).replace("BYTES:", "");
+            let json_msg: Value = serde_json::from_str(&cleaned_json_string.to_string()).expect("issue in from str"); // Assuming serde_json is used for JSON serialization/deserialization
+            let action = json_msg["action"].clone();
+            let json_data_value = json_msg["data"].clone();
+            
+            let json_str_ref = action.as_str().expect("issue in converting value to string");
+
+            let status = match json_str_ref {
+                SEND_COMMAND => {
+                    let wu_message: WrapperUMessage =
+                        serde_json::from_value(json_data_value).unwrap(); // convert json to UMessage
+                    let  u_message = wu_message.0;
+                    action_str = constants::SEND_COMMAND;
+                    self.utransport.send(u_message).await
+                }
+
+                REGISTER_LISTENER_COMMAND => {
+                    let cloned_listener = Arc::clone(&arc_self);
```

Comment:
So every listener registered is desired to be the same one?

That's fine. I'd recommend `SocketTestAgent` hold onto an `Arc<Mutex<SocketTestListener>>` then and then you can clone it before entering the loop.

Reply:
Waiting to hear back on this...